File: rs/src/key.rs
//! Key type with lifecycle tracking.

use coz::Thumbprint;

/// Revocation information for a key.
#[derive(Debug, Clone)]
pub struct Revocation {
    /// Revocation timestamp (SPEC `rvk` field).
    pub rvk: i64,
    /// Thumbprint of key that performed revocation (None = self-revoke).
    pub by: Option<Thumbprint>,
}

/// Cyphrpass key with lifecycle tracking.
///
/// Extends the Coz key concept with Cyphrpass-specific fields for
/// tracking when keys were added, last used, and revoked.
#[derive(Debug, Clone)]
pub struct Key {
    /// Algorithm identifier (e.g., "ES256").
    pub alg: String,
    /// Key thumbprint.
    pub tmb: Thumbprint,
    /// Public key bytes.
    pub pub_key: Vec<u8>,
    /// When this key was first added to the principal.
    pub first_seen: i64,
    /// When this key last signed a valid transaction/action.
    pub last_used: Option<i64>,
    /// Revocation info (None = active).
    pub revocation: Option<Revocation>,
    /// Optional human-readable label.
    pub tag: Option<String>,
}

impl Key {
    /// Returns true if the key is currently active (not revoked).
    pub fn is_active(&self) -> bool {
        self.revocation.is_none()
    }

    /// Returns true if the key was active at the given timestamp.
    ///
    /// Critical for validating historical transactions.
    pub fn is_active_at(&self, timestamp: i64) -> bool {
        self.revocation.as_ref().map_or(true, |r| timestamp < r.rvk)
    }
}

File: go/key.go
package cyphr

import "github.com/cyphrme/coz"

// Revocation holds information about a revoked key.
type Revocation struct {
	// Rvk is the revocation timestamp (SPEC rvk field).
	// Signatures with now >= rvk are invalid.
	Rvk int64

	// By is the thumbprint of the key that performed the revocation.
	// nil for self-revoke, non-nil for other-revoke.
	By *coz.B64
}

// Key extends coz.Key with Cyphr lifecycle tracking.
type Key struct {
	*coz.Key

	// FirstSeen is when this key was first added to the principal.
	FirstSeen int64

	// LastUsed is the last time this key signed a valid coz/action.
	LastUsed int64

	// Revocation holds revocation info. nil if the key is active.
	Revocation *Revocation
}

// IsActive returns true if the key is not revoked.
func (k *Key) IsActive() bool {
	return k.Revocation == nil
}

// IsActiveAt returns true if the key was active at the given timestamp.
// This is critical for validating historical cozies.
// Per SPEC §11.3: Signatures with now >= rvk are invalid.
func (k *Key) IsActiveAt(timestamp int64) bool {
	if k.Revocation == nil {
		return true
	}
	return timestamp < k.Revocation.Rvk
}
